add readme for smart dialer JSON config (#302)
Jigsaw-Code/outline-sdk: x/smart/README.md

# Smart Dialer

> 

Reviewer: We need to explain the Smart Dialer first, before we dive into the config.

The **Smart Dialer** searches for a strategy that unblocks DNS and TLS for a given list of test domains. It takes a config describing multiple strategies to pick from.

## JSON config for the Smart Dialer

The config that the Smart Dialer takes is in a JSON format. Here is an example:

```json

> {
>   "dns": [
>     {
>       "https": {
>         "name": "doh.sb"
>       }
>     }
>   ],
>   "tls": [
>     "override:host=cloudflare.net|tlsfrag:1"
>   ]
> }
> 

Reviewer: I think it would be clearer if use use multiple options. Also, we should have no circumvention as the first option. I'd I'd rather avoid `dns.sb` which is more obscure.

```suggestion
{
  "dns": [
      {"system": {}},
      {"https": {"name": "8.8.8.8"}},
      {"https": {"name": "9.9.9.9"}}
  ],
  "tls": [
      "",
      "split:2",
      "tlsfrag:1"
  ]
}
```

```

### DNS Configuration

*   The `dns` field specifies a list of DNS resolvers to test.

> *   Each DNS resolver can be one of the following types:

Reviewer: You can't. This specifies the type, they are mutually exclusive.

    *   `system`: Use the system resolver. Specify with an empty object.
    *   `https`: Use an encrypted DNS over HTTPS (DoH) resolver.
    *   `tls`: Use an encrypted DNS over TLS (DoT) resolver.
    *   `udp`: Use a UDP resolver.
    *   `tcp`: Use a TCP resolver.

#### DNS-over-HTTPS Resolver (DoH)

```json

> {
>   "https": {
>     "name": "doh.sb",
>     "address": "doh.sb:443"
>   }
> }

Reviewer: The address in the given example is not doing anything. Perhaps:

```suggestion
{
  "https": {
    "name": "dns.google",
    "address": "8.8.8.8"
  }
}
```

```

*   `name`: The domain name of the DoH server.
*   `address`: The host:port of the DoH server. Defaults to `name`:443.

#### DNS-over-TLS Resolver (DoT)

> ```json
> {
>   "tls": {
>     "name": "dns.google",
>     "address": "dns.google:853"
>   }
> }
> 
> ```

Reviewer: ```suggestion
```json
{
  "tls": {
    "name": "dns.google",
    "address": "8.8.8.8"
  }
}
```
```

*   `name`: The domain name of the DoT server.
*   `address`: The host:port of the DoT server. Defaults to `name`:853.

#### UDP Resolver

```json
{
  "udp": {

>     "address": "8.8.8.8:53"

Reviewer: Port 53 is the default
```suggestion
    "address": "8.8.8.8"
```

  }
}

```

*   `address`: The host:port of the UDP resolver.

#### TCP Resolver

> ```json
> {
>   "tcp": {
>     "address": "1.1.1.1:53"
>   }
> }
> 
> ```

Reviewer: Port 53 is the default
```suggestion
```json
{
  "tcp": {
    "address": "8.8.8.8"
  }
}
```
```

*   `address`: The host:port of the TCP resolver.

### TLS Configuration

*   The `tls` field specifies a list of TLS transports to test.
*   Each TLS transport is a string that specifies the transport to use.

> *   For example, `override:host=cloudflare.net|tlsfrag:1` specifies a transport that uses domain fronting with Cloudflare and TLS fragmentation. The available transports are defined in the `outline-sdk/x/configurl` package.

Reviewer: https://pkg.go.dev/github.com/Jigsaw-Code/outline-sdk/x/config#hdr-Config_Format

### Using the Smart Dialer

To use the Smart Dialer, create a `StrategyFinder` object and call the `NewDialer` method, passing in the list of test domains and the JSON config. The `NewDialer` method will return a `transport.StreamDialer` that can be used to create connections using the found strategy. For example:

```go
finder := &smart.StrategyFinder{
    TestTimeout:  5 * time.Second,
    LogWriter:   os.Stdout,
    StreamDialer: &transport.TCPDialer{},
    PacketDialer: &transport.UDPDialer{},
}

configBytes := []byte(`

> {
>   "dns": [
>     {
>       "https": {
>         "name": "doh.sb"
>       }
>     }
>   ],
>   "tls": [
>     "override:host=cloudflare.net|tlsfrag:1"
>   ]
> }

Reviewer: Using the initial example:

```suggestion
{
  "dns": [
      {"system": {}},
      {"https": {"name": "8.8.8.8"}},
      {"https": {"name": "9.9.9.9"}}
  ],
  "tls": [
      "",
      "split:2",
      "tlsfrag:1"
  ]
}
```

# Smart Dialer

> The **Smart Dialer** is a dialer that **automatically chooses the best way to connect** to the internet from a list of configured options.

Reviewer: "Best" is not really well defined, and we don't really do that.
We just pick a way that ensures we can successfully connect to a list of given services.

I think saying "connect to the internet" also conveys something else.

> The **Smart Dialer** is a dialer that **automatically chooses the best way to connect** to the internet from a list of configured options.

Reviewer: ```suggestion
The **Smart Dialer** searches for a strategy that unblocks DNS and TLS for a given list of test domains. It takes a config describing multiple strategies, which is used to pick from.
```

Reviewer: Maybe something like this, to borrow from the text below (which should probably be updated to remove the duplication).

## JSON config for the Smart Dialer

The Smart Dialer uses a JSON config to dynamically find serverless strategies for circumvention. The config is used to search for a strategy that unblocks DNS and TLS for a given list of test domains.

Here is an example of a JSON config:

```json
{
  "dns": [
    {
      "https": {
        "name": "doh.sb"
      }
    }
  ],
  "tls": [
    "override:host=cloudflare.net|tlsfrag:1"
  ]
}

```

### DNS Configuration

*   The `dns` field specifies a list of DNS resolvers to test.
*   Each DNS resolver can be one of the following types:
    *   `system`: Use the system resolver. Specify with an empty object.
    *   `https`: Use an encrypted DNS over HTTPS (DoH) resolver.
    *   `tls`: Use an encrypted DNS over TLS (DoT) resolver.
    *   `udp`: Use a UDP resolver.
    *   `tcp`: Use a TCP resolver.

> #### HTTPS Resolver

Reviewer: ```suggestion
#### DNS-over-HTTPS Resolver (DoH)
```

```json
{
  "https": {
    "name": "doh.sb",
    "address": "doh.sb:443"
  }
}

```

*   `name`: The domain name of the DoH server.
*   `address`: The host:port of the DoH server. Defaults to `name`:443.

> #### TLS Resolver

Reviewer: ```suggestion
#### DNS-over-TLS Resolver (DoT)
```

```json
{
  "tls": {
    "name": "dns.google",
    "address": "dns.google:853"
  }
}

```

*   `name`: The domain name of the DoT server.
*   `address`: The host:port of the DoT server. Defaults to `name`:853.

#### UDP Resolver

```json
{
  "udp": {
    "address": "8.8.8.8:53"
  }
}

```

*   `address`: The host:port of the UDP resolver.

#### TCP Resolver

```json
{
  "tcp": {
    "address": "1.1.1.1:53"
  }
}

```

*   `address`: The host:port of the TCP resolver.

### TLS Configuration

*   The `tls` field specifies a list of TLS transports to test.
*   Each TLS transport is a string that specifies the transport to use.

> *   For example, `override:host=cloudflare.net|tlsfrag:1` specifies a transport that uses domain fronting with Cloudflare and TLS fragmentation. [The available transports are documented here.](https://pkg.go.dev/github.com/Jigsaw-Code/outline-sdk/x/config#hdr-Config_Format)

Reviewer: This is clearer hyperlinking:

```suggestion
*   For example, `override:host=cloudflare.net|tlsfrag:1` specifies a transport that uses domain fronting with Cloudflare and TLS fragmentation. See the [config documentation](https://pkg.go.dev/github.com/Jigsaw-Code/outline-sdk/x/config#hdr-Config_Format) for details.
```
